transactions/WFLY-12610-transaction-recovery-process-operation.adoc

```diff
@@ -0,0 +1,137 @@
+to speed up resolving the prepared unfinished transactions from the Narayana object store.
+
+== Issue Metadata
+
+=== Issue
+
+* https://issues.redhat.com/browse/WFLY-12610[WFLY-12610]
+
+=== Related Issues
+
+* https://issues.redhat.com/browse/JBEAP-20714[JBEAP-20714]
+
+=== Dev Contacts
+
+* mailto:ochaloup@redhat.com[{author}]
+
+=== QE Contacts
+
+* mailto: TBD
+
+=== Testing By
+
+* [x] Engineering
+* [ ] QE
+
+=== Affected Projects or Components
+
+* Transactions
+
+=== Other Interested Projects
+
+* WildFly Operator
+
+=== Relevant Installation Types
+
+* [x] Traditional standalone server (unzipped or provisioned by Galleon)
+* [ ] Managed domain
+* [ ] OpenShift s2i
+* [ ] Bootable jar
+
+== Motivation
+
+These operations will be used for OpenShift scaledown processing. The current processing is run
+https://github.com/wildfly/wildfly-operator/blob/0.5.1/pkg/controller/wildflyserver/transaction_recovery.go#L196[by WildFly Operator].
+The WildFly scaledown processing needs to ensure that all transactions are finished before the pod is removed from the cluster.
+With the pod removal the object store content is lost. If the Narayana object store contained some in-doubt transaction then it is never restored.
+The notion about it is removed at time the pod is removed.
+For not losing the transactions the scaledown processing removes all unfinished transactions before it permits the pod to be removed from the OpenShift.
+Resolving the in-doubt transaction is the task for recovery manager. The CLI operations, that are added by this proposal, provide a way to manually
+launch the recovery processing and set up the Narayana transaction manager in way to speed-up the recovery to successfully resolve the in-doubt records.
+
+Until now the configuration has to be done with use of the system properties, enabling the special network socket and restarting JVM.
+All these operations are time consuming and do not permit the recovery to be processed with https://issues.redhat.com/browse/JBEAP-20714[bootable jar].
+
+== Requirements
+
+Ability to manually launch the transaction recovery and configure the behaviour
+with changing the fundamental attributes via JBoss CLI.
+
+=== Hard Requirements
+
+Adding the following operations and configuration attributes to transaction subsystem to XML definition and as the JBoss CLI operations
+
+[cols="1,1,1"]
+|===
+|Name |Type |Description
+
+|process-recovery
+|operation
+|Execute the transaction recovery processing. Runtime operation.
+
+|orphan-safety-interval
+|attribute
+|Defines a time after which the record at remote resource without counterpart in the object store is considered as orphan and rolled-back, in milliseconds. Runtime, no reload needed.
+
+|transactions-enabled
+|attribute
+|When true a new transaction can be started, when false starting a new transaction is forbidden. Runtime, no reload needed.
```

Comment:
I would be wary of adding a `transactions-enabled` operation since that could be abused or inadvertently configured. If you embed or move the functionality into another operation that is sensitive to the starting of new transactions then that would be safer. Is there another operation where that could be done?

Otherwise the feature looks useful.

Reply:
Thanks for the feedback. I see, having the special attribute to disable starting transaction could not be appropriate in all contexts.
I would propose to have a chance to disable starting new transactions when the server goes suspended - i.e. adding a new attribute to `suspend` operation, like `:suspend(transactions-enabled=false)`

Reply:
@mmusgrov I discussed the possibilities of the `:suspend` operation with @kabir and I found that the `:suspend` operation is a generic WF core operation that is not permitted to be used directly from WildFly subsystems.
I would use a similar way of having a special attribute which activates the action only on suspend.
It would mean to rename the attribute to `transactions-enabled-on-suspend` (would you have a better name?) with value `true/false`. The functionality will be to disable transactions only when `:suspend` is called and enabled when container is resumed.
What do you think about such approach?

Reply:
I think the approach is good (and the attribute name describes its intent).